File: BAEKJOON/카테고리/스택/4949_균형잡힌_세상/4949.py
import sys

sys.stdin = open('input.txt', 'r')

while True:
    sentence = input()

    if sentence == '.':
        break

    stack = []

    result = 'yes'

    for char in sentence:
        if char == '(' or char == '[':
            stack.append(char)

        elif char == ']':
            if len(stack) == 0:
                result = 'no'
                break

            popped = stack.pop()

            if popped != '[':
                result = 'no'

        elif char == ')':
            if len(stack) == 0:
                result = 'no'
                break

            popped = stack.pop()

            if popped != '(':
                result = 'no'
                break

    if len(stack) != 0:
        result = 'no'

    print(result)

File: BAEKJOON/4949_균형잡힌세상/4949.js
const fs = require('fs');
const fileStream = fs.createReadStream('input.txt');
const readline = require('readline');

const rl = readline.createInterface({
  input: fileStream,
  output: process.stdout,
})

let input = []

function isBeautifulWorld(sentence) {
  const stack = [];
  const pair = {
    "[": "]",
    "(": ")",
  }
  for (let i = 0; i < sentence.length; i++) {
    const char = sentence[i];
    if (char === '(' || char === '[') {
      stack.push(char)
    } else if (char === ')' || char === ']') {
      if (stack.length > 0 && pair[stack[stack.length - 1]] === char) {
        stack.pop()
      } else {
        return 'no'
      }
    }
  }
  return stack.length > 0 ? 'no' : 'yes';
}

rl.on("line", (line) => {
  input.push(line);
})
  .on("close", () => {
    input.slice(0,input.length-1).forEach(sentence => console.log(isBeautifulWorld(sentence)));
    process.exit()
  })